pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,346 @@
+
+package metrics
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/pkg/testutil"
+)
+
+const (
+	testAppID     = "asdf"
+	testVersion   = "1.0.0"
+	testInstallID = "yv66vt6tvu8="
+	testServerURL = "https://example.com"
+)
+
+func defaultClient() Client {
+	return Client{
+		appID:      testAppID,
+		version:    testVersion,
+		installID:  testInstallID,
+		httpClient: &http.Client{Timeout: 1 * time.Second},
+		optOut:     false,
+		config: &metricsConfig{
+			ServerURL: testServerURL,
+			NoMetrics: false,
+		},
+	}
+}
+
+// Not all failure cases can be easily tested, will test subset that is easy
+// to reproduce.
+func Test_New_unhappy(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		name      string
+		appID     string
+		env       map[string]string
+		wantError string
+	}{
+		{
+			name:      "empty_app_id_noop",
+			appID:     "",
+			wantError: "appID cannot be empty",
+		},
+		{
+			name:      "opt_out_env_noop_no_err",
+			appID:     testAppID,
+			env:       map[string]string{"NO_METRICS": "TRUE"},
+			wantError: "",
+		},
+		{
+			name:      "bad_url_noop",
+			appID:     testAppID,
+			env:       map[string]string{"METRICS_URL": "htttpq://%foo*(*fg.com4/\\"},
+			wantError: "failed to parse server URL",
+		},
+	}
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			client, err := New(ctx, tc.appID, "1", WithLookuper(envconfig.MapLookuper(tc.env)))
+			if diff := testutil.DiffErrString(err, tc.wantError); len(diff) != 0 {
+				t.Errorf("unexpected error: %s", diff)
+			}
+			if client == nil {
+				t.Errorf("client returned should be non-nil")
+			}
+			if !client.optOut {
+				t.Errorf("client returned should be optOut")
+			}
+		})
+	}
+}
+
+func Test_New_Happy(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name      string
+		client    *http.Client
+		installID string
+		want      Client
+	}{
+		{
+			name: "happy_path_no_install_id",
+			want: defaultClient(),
+		},
+		{
+			name:      "happy_path_with_install_id",
+			installID: testInstallID,
+			want:      defaultClient(),
+		},
+		{
+			name:      "happy_path_with_custom_http_client",
+			installID: testInstallID,
+			client:    &http.Client{Timeout: 2},
+			want: func() Client {
+				c := defaultClient()
+				c.httpClient = &http.Client{Timeout: 2}
+				return c
+			}(),
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+
+			installPath := t.TempDir() + "/" + installIDFileName
+			if len(tc.installID) > 0 {
+				if err := storeInstallID(testAppID, installPath, &InstallIDData{tc.installID}); err != nil {
+					t.Fatalf("test setup failed: %s", err.Error())
+				}
+			}
```

Comment:
What is this guard for? Also it's more idiomatic to check `tc.installID != ""`.

Reply:
The zero value of installID in test is stating we don't want any existing install id file. This guard is so we only attempt to store in cases of a non-zero-value id.